File: d1/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader, Read, Seek, SeekFrom};

fn main() {
    let file = File::open("d1-1.txt").unwrap();
    let mut reader = BufReader::new(file);

    let mut part1_sum = 0;
    for line in reader.by_ref().lines() {
        let digits: Vec<_> = line.unwrap()
            .chars()
            .filter(|c| c.is_ascii_digit())
            .map(|c| c.to_digit(10).unwrap())
            .collect();
        part1_sum += match digits.len() {
            1 => digits[0] * 10 + digits[0],
            _ => digits[0] * 10 + digits[digits.len() - 1],
        }
    }

    println!("part 1 sum: {part1_sum}");

    reader.seek(SeekFrom::Start(0)).unwrap();

    let digit_strings = [
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
    ];
    let digit_map: HashMap<&&str, u32> = HashMap::from_iter(digit_strings.iter()
        .enumerate()
        .map(|(i, k)| (k, i as u32 + 1)));

    let mut part2_sum = 0;
    for line in reader.lines().map(|line| line.unwrap()) {
        let mut digits: Vec<_> = digit_strings.iter()
            .map(|s| (line.match_indices(s), *digit_map.get(s).unwrap() as u32))
            .flat_map(|(i, e)| i.map(move |ii| (ii, e)))
            .collect();
        digits.extend(
            (0..10).into_iter()
                .map(|i| (line.match_indices(char::from_digit(i, 10).unwrap()), i))
                .flat_map(|(i, e)| i.map(move |ii| (ii, e)))
        );
        digits.sort_by_key(|(i, _)| i.0);
        part2_sum += match digits.len() {
            1 => digits[0].1 * 10 + digits[0].1,
            _ => digits[0].1 * 10 + digits[digits.len() - 1].1,
        }
    }
    println!("part 2 sum: {part2_sum}");
}

File: d1/src/main.cpp
#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <vector>

std::map<std::string, int> get_str_map() {
    const std::string digit_strings[] = {
            "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
    std::map<std::string, int> str_to_digit;
    int i = 1;
    for (auto s: digit_strings) {
        str_to_digit[s] = i++;
    }
    return str_to_digit;
}

int main() {
    std::ifstream file("../d1-1.txt");

    if (!file.is_open()) {
        throw;
    }

    long part1_sum = 0;
    long part2_sum = 0;

    const std::map<std::string, int> str_to_digit = get_str_map();


    std::string line;
    while (std::getline(file, line)) {
        std::vector<int> digits;
        std::vector<std::pair<int, int>> digit_positions;

        int i = 0;
        for (const auto c: line) {
            const auto digit = c - '0';
            if (digit >= 0 && digit <= 9) {
                digits.push_back(digit);
                digit_positions.emplace_back(i, digit);
            }
            i++;
        }
        for (const auto&[digit_str, digit]: str_to_digit) {
            auto pos = line.find(digit_str, 0);
            while (pos != std::string::npos) {
                digit_positions.emplace_back(pos, digit);
                pos = line.find(digit_str, pos + 1);
            }
        }

        assert(!digits.empty());
        part1_sum += digits[0] * 10 + digits[digits.size() - 1];

        std::sort(digit_positions.begin(), digit_positions.end());
        part2_sum += digit_positions[0].second * 10 + digit_positions[digit_positions.size() - 1].second;
    }
    file.close();

    std::cout << "Part 1 sum: " << part1_sum << std::endl;
    std::cout << "Part 2 sum: " << part2_sum << std::endl;
}
